File: fleet/util/os/src/fleet/util/os/Os.java
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package fleet.util.os;

import java.util.Locale;

/**
 * @deprecated Use {@link fleet.util.Os} instead.
 */
@Deprecated
public final class Os {
  // all these fields are for compatibility, we may replace them with something more manageable
  public static final Os INSTANCE = new Os();

  private Os() {
  }

  public enum Type {
    Windows, Linux, MacOS, Unknown
  }

  public String getName() {
    return System.getProperty("os.name");
  }

  public String getVersion() {
    return System.getProperty("os.version");
  }

  public String getArch() {
    return System.getProperty("os.arch");
  }

  public Type getType() {
    String normalizedName = getName().toLowerCase(Locale.ROOT);
    if (normalizedName.startsWith("mac")) {
      return Type.MacOS;
    }
    if (normalizedName.startsWith("win")) {
      return Type.Windows;
    }
    if (normalizedName.contains("nix") || normalizedName.contains("nux")) {
      return Type.Linux;
    }
    return Type.Unknown;
  }

  public boolean isAarch64() {
    String arch = getArch();
    return "aarch64".equals(arch) || "arm64".equals(arch);
  }

  public boolean isMac() {
    return getType() == Type.MacOS;
  }

  public boolean isWindows() {
    return getType() == Type.Windows;
  }

  public boolean isLinux() {
    return getType() == Type.Linux;
  }

  public boolean isUnix() {
    Type type = getType();
    return type == Type.Linux || type == Type.MacOS;
  }
}

File: fleet/util/core/srcCommonMain/fleet/util/Os.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package fleet.util

import fleet.util.multiplatform.linkToActual

class Os private constructor() {
  enum class Type {
    Windows, Linux, MacOS, Unknown
  }

  val name: String
    get() = getName()

  val version: String
    get() = getVersion()

  val arch: String
    get() = getArch()

  val type: Type
    get() {
      val normalizedName = this.name.lowercase()
      return when {
        normalizedName.startsWith("mac") -> Type.MacOS
        normalizedName.startsWith("win") -> Type.Windows
        normalizedName.contains("nix") || normalizedName.contains("nux") -> Type.Linux
        else -> Type.Unknown
      }
    }

  val isAarch64: Boolean
    get() {
      val arch = this.arch
      return "aarch64" == arch || "arm64" == arch
    }

  val isWasm: Boolean
    get() {
      val arch = this.arch
      return "wasm" == arch
    }

  val isMac: Boolean
    get() = this.type == Type.MacOS

  val isWindows: Boolean
    get() = this.type == Type.Windows

  val isLinux: Boolean
    get() = this.type == Type.Linux

  val isUnix: Boolean
    get() {
      val type = this.type
      return type == Type.Linux || type == Type.MacOS
    }

  companion object {
    val INSTANCE: Os = Os()
  }
}

internal fun getName(): String = linkToActual()

internal fun getVersion(): String = linkToActual()

internal fun getArch(): String = linkToActual()
